### PR TITLE
show war packages in sub-tree in Tomcat Server explorer

### PR DESCRIPTION
**Pull request by SummerSun**

### CODE REVIEW

File: src/Tomcat/TomcatController.ts [hunk 1]
```diff
@@ -230,34 +242,22 @@ export class TomcatController {
...
-                statusBarCommand = vscode.commands.registerCommand(statusBar.command, async () => {
-                    opn(serverUri);
-                });
-                statusBar.show();
-            }
-
             watcher = chokidar.watch(serverConfig);
             watcher.on('change', async () => {
                 if (serverPort !== await Utility.getPort(serverConfig, Constants.PortKind.Server)) {
                     vscode.window.showErrorMessage(localize('tomcatExt.serverPortChangeError', `Changing the server port of a running server will cause errors, please change it back to ${ serverPort} ！`));
```
Comment (jdneo) [inline src/Tomcat/TomcatController.ts]: Looks like we have a wrong `！`here

File: src/TomcatSeverTreeProvider.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+export class TomcatSeverTreeProvider implements vscode.TreeDataProvider<TreeItem> {
+    public _onDidChangeTreeData: vscode.EventEmitter<TreeItem> = new vscode.EventEmitter<TreeItem>();
+    public readonly onDidChangeTreeData: vscode.Event<TreeItem> = this._onDidChangeTreeData.event;
+
+    constructor(private _context: vscode.ExtensionContext, private _tomcatModel: TomcatModel) {
+        this._onDidChangeTreeData.fire();
+    }
+
+    public async getTreeItem(element: TreeItem): Promise<TreeItem> {
+        /* qisun: checking http port is not a valid way to detect the server running or not
```
Comment (jdneo) [inline src/TomcatSeverTreeProvider.ts]: remove the comments if they are not used anymore

File: src/Tomcat/TomcatController.ts (hunk 1 above)
Comment (SummerSun) [reply src/Tomcat/TomcatController.ts]: the prompt message was moved to DialogMessage in another PR, will update the branch after another PR merged.

File: package.json [hunk 3]
```diff
@@ -185,16 +189,20 @@
...
           "command": "tomcat.server.delete",
           "when": "view == tomcatServerExplorer && viewItem == runningserver",
-          "group": "tomcat@5"
+          "group": "tomcat@6"
         },
         {
           "command": "tomcat.server.delete",
           "when": "view == tomcatServerExplorer && viewItem == idleserver",
-          "group": "tomcat@5"
+          "group": "tomcat@6"
```
Comment (jdneo) [inline package.json]: two `tomcat@6` here?

File: src/Tomcat/TomcatController.ts [hunk 4]
```diff
@@ -57,6 +58,17 @@ export class TomcatController {
         vscode.window.showTextDocument(vscode.Uri.file(configFile), { preview: false });
     }
 
+    public async browseWarPackage(warPackage: WarPackage): Promise<void> {
+        if (warPackage) {
+            const server: TomcatServer = this._tomcatModel.getTomcatServer(warPackage.serverName);
+            opn(await this.getServerUri(server, warPackage.label));
+            if (!server.isStarted()) {
+                server.outputChannel.appendLine('Start server to see real war package');
```
Comment (akaroml) [inline src/Tomcat/TomcatController.ts]: Why not prompt the user to start the server using actions?

File: src/Tomcat/TomcatController.ts [hunk 5]
```diff
@@ -247,28 +259,14 @@ export class TomcatController {
     }
 
     private async startTomcat(serverInfo: TomcatServer, appName: string): Promise<void> {
-        let statusBar: vscode.StatusBarItem;
```
Comment (akaroml) [inline src/Tomcat/TomcatController.ts]: As we delete these lines, where does the logic go?

File: src/TomcatSeverTreeProvider.ts (hunk 2 above)
Comment (akaroml) [reply src/TomcatSeverTreeProvider.ts]: Agreed, just delete whatever is outdated.

File: src/Tomcat/TomcatController.ts (hunk 4 above)
Comment (SummerSun) [reply src/Tomcat/TomcatController.ts]: good idea

File: resources/war.svg [hunk 6]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
```
[the remark this answers is not in the document]
Comment (SummerSun) [reply resources/war.svg]: yeah, make sense, will remove the image for now

File: src/Tomcat/TomcatController.ts (hunk 5 above)
Comment (SummerSun) [reply src/Tomcat/TomcatController.ts]: status bar command was used to open war package url in browser in previous behavior. Now we are supporting showing war packages directly in sub-tree and a context command to "open in browser".

**Comment by SummerSun** [comment]
#90

**Pull request merged**